### PR TITLE
Add text alignment

### PR DESCRIPTION
**Pull request by zjp**
Let's try this again with a clean git log -- sorry about that. 

## Description

Adds the option to align the output to the space after the end of the separator string ":", "->", etc by padding all arguments to printf by the same amount regardless of length 

## Features
--align_output 'left', 'right', 'off'
left: Don't right-justify labels like "OS", but align what gets listed after
right: Align labels to make their contents aligned
off: Default behavior of neofetch

## Issues

1. Addresses #1241 and #895 

Screenshot 2019-07-12 at 16 46 11
Screenshot 2019-07-12 at 16 46 28
Screenshot 2019-07-12 at 16 46 49

### CODE REVIEW

File: neofetch [hunk 1]
```diff
@@ -3978,27 +3984,31 @@ prin() {
...
         string="${2:-$1}"
         local subtitle_color="$info_color"
     fi
-
     string="$(trim "${string//$'\e[0m'}")"
     length="$(strip_sequences "$string")"
     length="${#length}"
 
-    # Format the output.
+    # # Format the output.
```
Comment (dylanaraps) [inline neofetch]: This can be removed.

File: neofetch [hunk 2]
```diff
@@ -3978,27 +3984,31 @@ prin() {
     if [[ "$(trim "$1")" && "$2" ]]; then
         [[ "$json" ]] && { printf '    %s\n' "\"${1}\": \"${2}\","; return; }
 
-        string="${1}${2:+: $2}"
+        string="${1}"
```
Comment (dylanaraps) [inline neofetch]: Why was this removed?

File: neofetch [hunk 3]
```diff
@@ -3978,27 +3984,31 @@ prin() {
...
+    printf '%b' "${text_padding:+\e[${text_padding}C}${zws}"
+    case "$align_output" in
+        "left")  printf '%-0*b' "$offset" "${string}${separator} " ;;
+        "right") printf '%+0*b' "$offset" "${string}${separator} " ;;
+        "off") printf '%b' "${string}${separator} " ;;
+    esac
+    printf '%b\n' "${reset}${2}"
 
-    # Calculate info height.
+    # # Calculate info height.
```
Comment (dylanaraps) [inline neofetch]: This can be removed.

File: neofetch [hunk 4]
```diff
@@ -3978,27 +3984,31 @@ prin() {
...
     ((++info_height))
 
-    # Log that prin was used.
+    # # Log that prin was used.
```
Comment (dylanaraps) [inline neofetch]: This can be removed.

File: neofetch [hunk 5]
```diff
@@ -4645,7 +4656,8 @@ get_args() {
             "--underline") underline_enabled="$2" ;;
             "--underline_char") underline_char="$2" ;;
             "--bold") bold="$2" ;;
-            "--separator") separator="$2" ;;
+            "--separator") separator="$2";;
```
Comment (dylanaraps) [inline neofetch]: Space between `"` and `;;`.

File: neofetch [hunk 6]
```diff
@@ -4645,7 +4656,8 @@ get_args() {
...
+            "--align_output") align_output="$2";;
```
Comment (dylanaraps) [inline neofetch]: Space between `"` and `;;`.

File: neofetch [hunk 7]
```diff
@@ -8968,6 +8980,16 @@ EOF
     fi
 }
 
+get_offset() {
+#   longest_string
```
Comment (dylanaraps) [inline neofetch]:
```sh
    # Longest string.
```

File: neofetch (hunk 2 above)
Comment (zjp) [reply neofetch]: The way things are printed in this PR makes concatenating arg2 to arg1 redundant, but arg1 is still formatted correctly.

**Comment by zjp** [comment]
Magic number for every label, as long as it's the longest label being output on that run 
```
     Label    |Length|Magic#|
--------------|------|------|
Terminal Font |  13  |  28  |
 Music Player |  12  |  27  |
   Resolution |  10  |  25  |
   GPU Driver |  10  |  25  |
    CPU Usage |   9  |  24  |
    Public IP |   9  |  24  |
     Packages |   8  |  23  |
     WM Theme |   8  |  23  |
     Terminal |   8  |  23  |
     Local IP |   8  |  23  |
      Battery |   7  |  22  |
       Kernel |   6  |  21  |
       Uptime |   6  |  21  |
       Memory |   6  |  21  |
       Locale |   6  |  21  |
        Shell |   5  |  20  |
        Theme |   5  |  20  |
        Icons |   5  |  20  |
        Users |   5  |  20  |
         Host |   4  |  19  |
         Disk |   4  |  19  |
         Font |   4  |  19  |
         Song |   4  |  19  |
          CPU |   3  |  18  |
          GPU |   3  |  18  |
           OS |   2  |  17  |
           DE |   2  |  17  |
           WM |   2  |  17  |

**Comment by zjp** [comment]
IT'S ALIVE! Finally. Pending your review, I think it's ready -- at least, it's as good as it can be given my novice bash-fu. 

Screen Shot 2019-11-16 at 7 16 52 AM

File: neofetch (hunk 2 above)
Comment (zjp) [reply neofetch]: I have a more complete explanation now. 

Originally, this if statement joined the two arguments into the same string, but to properly align the info by label, the labels and their corresponding data had to stay separate. I've left the JSON logic in place, but set string to the first argument in both cases, so I've updated the code to be 
```
# If $2 doesn't exist we format $1 as info.
if [[ "$(trim "$1")" && "$2" ]]; then
    [[ "$json" ]] && { printf '    %s\n' "\"${1}\": \"${2}\","; return; }
fi
string="${1}"
string="$(trim "${string//$'\e[0m'}")"
```

Now I'm not really sure what the lines 
```
length="$(strip_sequences "$string")"
length="${#length}"
```
do, but removing them doesn't seem to change anything, so I've left them in. 

Anyway, not concatenating $2 to $1 has several benefits:

First, we still strip all sequences from the labels, standardize them to bold, and get the length.

Second, now I can pad the first argument of calls to prin with an arbitrary number of `\033[0m  \033[0m  ...` and manually center text if I want to give my own short byline like this screenshot: 
Screenshot 2019-11-28 at 00 44 28

Third, we can still properly detect automatically whether or not the user probably wants a separator at the end of the string with 
```
string="${string}${reset}${colon_color}"
[[ "$2" ]] && string="${string}${separator}${separator:=:}"
```

Fourth: Now users can use arbitrary sequences to format text the way they like (italic, flashing, etc). 

With the string formatting simplified we can do:

```
printf '%b' "${text_padding:+\e[${text_padding}C}${zws}"
case "$align_output" in
    "left")  printf '%-*b' $offset "${string}" ;;
    "right") printf '%+*b' $offset "${string}" ;;
    "off") printf '%b' "${string}" ;;
esac
[[ "$2" ]] && printf '%b' "${info_color}${2}"
printf '\n'
```

If any of this breaks existing behavior I haven't witnessed it, so I'll need feedback from people with more exotic setups.

File: neofetch [hunk 8]
```diff
@@ -9663,14 +9710,20 @@ main() {
     [[ $TERM != minix && $stdout != on ]] && {
         # If the script exits for any reason, unhide the cursor.
         trap 'printf "\e[?25h\e[?7h"' EXIT
-
```
Comment (dylanaraps) [inline neofetch]: Add back this line.

File: neofetch [hunk 9]
```diff
@@ -9644,7 +9657,41 @@ EOF
...
+    # Longest string. I have no idea why this magic number 18
+    # exists, but it works for /every/ ascii art I tried.
+    longest_string_offset=$((18 + $(get_longest_label) + ${#separator}))
+    offset=$((text_padding - longest_string_offset))
+    offset=$((text_padding - offset))
+    echo $offset
+}
+
 main() {
+
```
Comment (dylanaraps) [inline neofetch]: Remove this line.

File: neofetch [hunk 10]
```diff
@@ -9663,14 +9710,20 @@ main() {
...
-
         # Hide the cursor and disable line wrap.
         printf '\e[?25l\e[?7l'
     }
 
     image_backend
     get_cache_dir
     old_functions
+
+    case "$align_output" in
```
Comment (dylanaraps) [inline neofetch]:
```sh
    case $align_output in
        left|right) offset=$(get_offset) 
    esac
```

File: neofetch [hunk 11]
```diff
@@ -9644,7 +9657,41 @@ EOF
...
+	echo "$longest_label_length"
+}
+
+get_offset() {
+    # Longest string. I have no idea why this magic number 18
+    # exists, but it works for /every/ ascii art I tried.
+    longest_string_offset=$((18 + $(get_longest_label) + ${#separator}))
+    offset=$((text_padding - longest_string_offset))
+    offset=$((text_padding - offset))
+    echo $offset
```
Comment (dylanaraps) [inline neofetch]: Remove this line.

File: neofetch [hunk 12]
```diff
@@ -5335,23 +5348,23 @@ EOF
...
+            ${c1}****${c2}_____${c1}***${c2}/${c1}*
+           ***${c2}/${c1}*******${c2}//${c1}***
+          **${c2}/${c1}********${c2}///${c1}*${c2}/${c1}**
+         **${c2}/${c1}*******${c2}////${c1}***${c2}/${c1}**
+        **${c2}/${c1}****${c2}//////.,${c1}****${c2}/${c1}**
+       ***${c2}/${c1}*****${c2}/////////${c1}**${c2}/${c1}***
+      ****${c2}/${c1}****    ${c2}/////${c1}***${c2}/${c1}****
+     ******${c2}/${c1}***  ${c2}////   ${c1}**${c2}/${c1}******
+    ********${c2}/${c1}* ${c2}///      ${c1}*${c2}/${c1}********
+  ,******     ${c2}// ______ /    ${c1}******,
```
Comment (dylanaraps) [inline neofetch]: Remove this change.

File: neofetch [hunk 13]
```diff
@@ -4557,6 +4568,7 @@ TEXT FORMATTING:
     --underline_char char       Character to use when underlining title
     --bold on/off               Enable/Disable bold text
     --separator string          Changes the default ':' separator to the specified string.
+    --align_output              Aligns the text output to the separator string
```
Comment (dylanaraps) [inline neofetch]: Maybe add (left|right|off) here somewhere (if possible).

File: neofetch [hunk 14]
```diff
@@ -4268,10 +4276,13 @@ get_user_config() {
...
     elif [[ -f "${XDG_CONFIG_HOME}/neofetch/config.conf" ]]; then
         source "${XDG_CONFIG_HOME}/neofetch/config.conf"
-        err "Config: Sourced user config.    (${XDG_CONFIG_HOME}/neofetch/config.conf)"
+		# We will need these assignments later to calculate the alignment offset
+        config_file="${XDG_CONFIG_HOME}/neofetch/config.conf"
+		err "Config: Sourced user config.    (${XDG_CONFIG_HOME}/neofetch/config.conf)"
 
     elif [[ -f "${XDG_CONFIG_HOME}/neofetch/config" ]]; then
         source "${XDG_CONFIG_HOME}/neofetch/config"
+		config_file="${XDG_CONFIG_HOME}/neofetch/config"
```
Comment (dylanaraps) [inline neofetch]: Indentation doesn't match.

File: neofetch [hunk 15]
```diff
@@ -4268,10 +4276,13 @@ get_user_config() {
 
     elif [[ -f "${XDG_CONFIG_HOME}/neofetch/config.conf" ]]; then
         source "${XDG_CONFIG_HOME}/neofetch/config.conf"
-        err "Config: Sourced user config.    (${XDG_CONFIG_HOME}/neofetch/config.conf)"
+		# We will need these assignments later to calculate the alignment offset
+        config_file="${XDG_CONFIG_HOME}/neofetch/config.conf"
+		err "Config: Sourced user config.    (${XDG_CONFIG_HOME}/neofetch/config.conf)"
```
Comment (dylanaraps) [inline neofetch]: Indentation doesn't match.

File: neofetch [hunk 16]
```diff
@@ -4268,10 +4276,13 @@ get_user_config() {
 
     elif [[ -f "${XDG_CONFIG_HOME}/neofetch/config.conf" ]]; then
         source "${XDG_CONFIG_HOME}/neofetch/config.conf"
-        err "Config: Sourced user config.    (${XDG_CONFIG_HOME}/neofetch/config.conf)"
+		# We will need these assignments later to calculate the alignment offset
```
Comment (dylanaraps) [inline neofetch]: Indentation doesn't match.

File: neofetch [hunk 17]
```diff
@@ -3230,7 +3236,6 @@ END
                     term_font=${term_font/xft:}
                     term_font=${term_font/:*}
                 ;;
-
```
Comment (dylanaraps) [inline neofetch]: Add back this line.

File: neofetch [hunk 18]
```diff
@@ -4083,23 +4088,26 @@ prin() {
...
     if [[ "$(trim "$1")" && "$2" ]]; then
         [[ "$json" ]] && { printf '    %s\n' "\"${1}\": \"${2}\","; return; }
-
-        string="${1}${2:+: $2}"
-    else
-        string="${2:-$1}"
-        local subtitle_color="$info_color"
     fi
 
+	string="${1}"
```
Comment (dylanaraps) [inline neofetch]: Indentation.

File: neofetch [hunk 19]
```diff
@@ -4083,23 +4088,26 @@ prin() {
     # If $2 doesn't exist we format $1 as info.
     if [[ "$(trim "$1")" && "$2" ]]; then
         [[ "$json" ]] && { printf '    %s\n' "\"${1}\": \"${2}\","; return; }
-
-        string="${1}${2:+: $2}"
-    else
-        string="${2:-$1}"
-        local subtitle_color="$info_color"
```
Comment (dylanaraps) [inline neofetch]: Remind me again why this was removed?

File: neofetch [hunk 20]
```diff
@@ -4083,23 +4088,26 @@ prin() {
...
     fi
 
+	string="${1}"
     string="$(trim "${string//$'\e[0m'}")"
     length="$(strip_sequences "$string")"
     length="${#length}"
 
     # Format the output.
-    string="${string/:/${reset}${colon_color}${separator:=:}${info_color}}"
+	string="${string/:/${reset}${colon_color}}"
```
Comment (dylanaraps) [inline neofetch]: Indentation.

File: neofetch (hunk 19 above)
Comment (zjp) [reply neofetch]: I have a more complete explanation now. 

Originally, this if statement joined the two arguments into the same string, but to properly align the info by label, the labels and their corresponding data had to stay separate. I've left the JSON logic in place, but set string to the first argument in both cases, so I've updated the code to be 
```
# If $2 doesn't exist we format $1 as info.
if [[ "$(trim "$1")" && "$2" ]]; then
    [[ "$json" ]] && { printf '    %s\n' "\"${1}\": \"${2}\","; return; }
fi
string="${1}"
string="$(trim "${string//$'\e[0m'}")"
```

Now I'm not really sure what the lines 
```
length="$(strip_sequences "$string")"
length="${#length}"
```
do, but removing them doesn't seem to change anything, so I've left them in. 

Anyway, not concatenating $2 to $1 has several benefits:

First, we still strip all sequences from the labels, standardize them to bold, and get the length.

Second, now I can pad the first argument of calls to prin with an arbitrary number of `\033[0m  \033[0m  ...` and manually center text if I want to give my own short byline like this screenshot: 
Screenshot 2019-11-28 at 00 44 28
BUT DO NOTE that using sequences in prin -- for whatever reason -- means the string is not considered by the alignment logic. For this statement to be true the sequences have to be raw (e.g. you have to put \33[0m in your string instead of ${reset}). If you try to use substitutions the whole strong is considered and it throws the info way to the right. I consider this a feature, but I can see how someone might want this to be fixed later, so that they can choose whether to exclude something and use ${reset}${bold}${italic} etc. 

Third, we can still properly detect automatically whether or not the user probably wants a separator at the end of the string with 
```
string="${string}${reset}${colon_color}"
[[ "$2" ]] && string="${string}${separator}${separator:=:}"
```

Fourth: Now users can use arbitrary sequences to format text the way they like (italic, flashing, etc). 

With the string formatting simplified we can do:

```
printf '%b' "${text_padding:+\e[${text_padding}C}${zws}"
case "$align_output" in
    "left")  printf '%-*b' $offset "${string}" ;;
    "right") printf '%+*b' $offset "${string}" ;;
    "off") printf '%b' "${string}" ;;
esac
[[ "$2" ]] && printf '%b' "${info_color}${2}"
printf '\n'
```

If any of this breaks existing behavior I haven't witnessed it, so I'll need feedback from people with more exotic setups.

File: neofetch (hunk 11 above)
Comment (zjp) [reply neofetch]: Can you clarify? This looks like the whole function -- did you mean the comment?

File: neofetch (hunk 12 above)
Comment (zjp) [reply neofetch]: I'm really not sure what my editor did here besides remove the trailing whitespace after the first asterisk -- it does print in the same format, and it looks like the rest of the spaces are still spaces. Weird.

File: neofetch (hunk 16 above)
Comment (zjp) [reply neofetch]: Fixed.

File: neofetch (hunk 17 above)
Comment (zjp) [reply neofetch]: Fixed.

File: neofetch (hunk 15 above)
Comment (zjp) [reply neofetch]: Fixed.

File: neofetch (hunk 14 above)
Comment (zjp) [reply neofetch]: Fixed.

File: neofetch (hunk 18 above)
Comment (zjp) [reply neofetch]: Fixed.

File: neofetch (hunk 15 above)
Comment (zjp) [reply neofetch]: Fixed.

File: neofetch (hunk 13 above)
Comment (zjp) [reply neofetch]: Done in new force-push.

File: neofetch (hunk 10 above)
Comment (zjp) [reply neofetch]: Done in new force-push.

File: neofetch (hunk 9 above)
Comment (zjp) [reply neofetch]: Fixed.

File: neofetch (hunk 8 above)
Comment (zjp) [reply neofetch]: Fixed.

File: neofetch (hunk 20 above)
Comment (zjp) [reply neofetch]: Fixed.

File: neofetch [hunk 21]
```diff
@@ -4083,23 +4089,26 @@ prin() {
...
     if [[ "$(trim "$1")" && "$2" ]]; then
         [[ "$json" ]] && { printf '    %s\n' "\"${1}\": \"${2}\","; return; }
-
-        string="${1}${2:+: $2}"
-    else
-        string="${2:-$1}"
-        local subtitle_color="$info_color"
     fi
 
+    string="${1}"
```
Comment (dylanaraps) [inline neofetch]:
```sh
string=$1
```

Comment (zjp) [reply neofetch]: Fixed in new force-push.

File: neofetch (hunk 11 above)
Comment (zjp) [reply neofetch]: I've removed the comment in a new force-push -- feel free to unresolve if necessary.

File: neofetch (hunk 12 above)
Comment (zjp) [reply neofetch]: I didn't think my initial solution of "it's the same so it's fine" was satisfactory, so I found a way to revert it cleanly. Fixed in the latest push -- sorry about that. It was a bit unprofessional of me in hindsight to suggest letting it be as a fix.

**Comment by zjp** [comment]
I went ahead and rebased this on master to see if it would fix those failing checks from a few months ago, and it seems that it has.
